sample adaptations

qwang98 opened the pull request:
<!--

Please follow this protocol when creating or reviewing PRs in this repository:

- Leave the PR as draft until review is required.
- When reviewing a PR, every reviewer should assign themselves as soon as they
  start, so that other reviewers know the PR is covered. You should not be
  discouraged from reviewing a PR with assignees, but you will know it is not
  strictly needed.
- Unless the PR is very small, help the reviewers by not making forced pushes, so
  that GitHub properly tracks what has been changed since the last review; use
  "merge" instead of "rebase". It can be squashed after approval.
- Once the comments have been addressed, explicitly let the reviewer know the PR
  is ready again.

-->

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -0,0 +1,95 @@
+let swap_u32_loop: int[25] -> int[25] = |st| array::map(array::new(25, |i| i), |i| swap_u32(st[i]));
+
+// ln 53 - 55
+let theta_bc: int[25], i -> int = |st, i| xor(st[i], xor(st[i + 5], xor(st[i + 10], xor(st[i + 15], st[i + 20]))));
```
To improve readability, we could also introduce a helper function
```suggestion
let xor_mult: int[] -> int = |input| std::array::fold(input, 0, |x, y| xor(x, y));
let theta_bc: int[25], i -> int = |st, i| xor_mult([st[i], st[i + 5], st[i + 10], st[i + 15], st[i + 20]]);
```

qwang98 replied:
Fixed

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -0,0 +1,95 @@
+
+// ln 53 - 55
+let theta_bc: int[25], i -> int = |st, i| xor(st[i], xor(st[i + 5], xor(st[i + 10], xor(st[i + 15], st[i + 20]))));
+let theta_bc_loop: int[25] -> int[5] = |st| array::map(array::new(5, |i| i), |i| theta_bc(st, i));
```
```suggestion
let theta_bc_loop: int[25] -> int[5] = |st| array::new(5, |i| theta_bc(st, i));
```

qwang98 replied:
Fixed

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -0,0 +1,142 @@
+};
+// rearrange st returned in ln 60 - version 2: needs to be called with theta_t_loop in another function
+let theta_t_loop_rearrange_v2: int[5][5] -> int[25] = |sp_partials| 
+    array::fold(array::new(5, |i| i), [], |initial, i| 
+        a + [sp_partials[0][i], sp_partials[1][i], sp_partials[2][i], sp_partials[3][i], sp_partials[4][i]]
+    );
```
```suggestion
    array::new(25, |i| sp_partials[i %5][i / 5]);
```

qwang98 replied:
Fixed

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -0,0 +1,142 @@
+    let st_partial = theta_st_loop(st, t, i); // int[5]
+    st_partial
+});
+// rearrange st returned in ln 60 - version 1: calls theta_t_loop
```
I don't understand, why do you need to rearrange?

Once the 5 bc values are computed in line 54, we can compute the new st values in line 60 in the following way

```
let theta = |bc, st| std::array::map_enumerated(st, |k, x| {
  let i = k % 5;
  let j = k / 5;
  let t = bc[(i + 4) % 5] ^ rotl64(bc[(i + 1) % 5], 1);
  xor(x, t)
});
```

qwang98 replied:
Agree on this. I just didn't know this kind of syntax for nested loop but now I do!

qwang98 commented:
Thanks for the very helpful comments. I finished a first draft of the keccak helper functions. With `main` as the entry function. A few questions:
1. The `else if` syntax doesn't seem to work for one instance I'm using it.
2. Should we test on 64 bit words to make sure the outputs match first, before adapting to 32 bit words?

Next steps will be constructing the `routine`, potentially flattening the circuit, and write up the main machine with PIL columns. Let me know if my understanding is correct.

chriseth commented:
Thanks a lot! I'll take another look.

1. The "else if" is not yet implemented, you have to use "else { if c { ... } }".
3. There is only the unlimited precision `int` type.
4. Yes, I think it's a good idea to have tests first, maybe even merge and then adapt to 32 bits.

2. and rest:  I think we should juts implement and test the keccakf function for now. It is not really clear yet how to pass teh input and output. Actually I think we can even remove the swap-part - since this only swaps endianness if I'm not mistaken.

I would say the next step is to test keccakf against a rust implementation (or also just some fixed example numbers).

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -0,0 +1,163 @@
+let theta_st: int[25] -> int[25] = |st| array::map_enumerated(st, |idx, elem| {
+    let i = idx % 5;
+    let j = idx / 5;
+    let t = xor(theta_bc(st, (i + 4) % 5), rotl32(theta_bc(st, (i + 1) % 5), 1));
```
Shouldn't this be rotl64?

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -0,0 +1,163 @@
+});
+// rearrange st_j
+let rho_pi_rearrange: int[25] -> int[25] = |st| {
+    let rearranged_st = [
+        st[23], st[17], st[5], st[11], st[6], st[22],
+        st[1], st[8], st[21], st[0], st[2], st[16], 
+        st[15], st[19], st[12], st[7], st[3], st[4], 
+        st[14], st[18], st[9], st[20], st[13], st[10]
+    ];
```
This is essentially the inverse of PI, right? Is there a way to compute it from PI?

qwang98 replied:
There's a functional way of doing this in a double loop, but I think it's not as efficient as just hard coding an inverse vector constant, which is what I did.

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -0,0 +1,163 @@
+    (new_st_j)
+};
+// collect st_j
+let rho_pi_loop: int[25] -> int[25] = |st| utils::fold(24, |i| i, [], |new_st, idx| {
+    let new_st_j = rho_pi(st, idx);
+    new_st + [new_st_j]
+});
```
Isn't this
```suggestion
let rho_pi_loop: int[25] -> int[25] = |st| std::array::new(25, |i| rho_pi(st, i));
});
```
?

qwang98 replied:
Agreed.

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -0,0 +1,163 @@
+// ln 74 - 83
+// chi
+// TODO: make sure that modulus has the same precedence as multiplication
+let chi: int[25] -> int[25] = |st| array::map_enumerated(st, |idx, elem| {
```
nice!

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -0,0 +1,163 @@
+let iota: int[25], int -> int[25] = |st, r| array::new(24, |i| if i == 0 { xor(st[0], RC[r]) } else { st[i] } );
+
+// ln 51 - 87
+let r_loop: int[25] -> int[25] = |st| utils::fold(24, |i| i, st, |acc, i| iota(chi(rho_pi_rearrange(rho_pi_loop(theta_st(acc)))), r) );
```
```suggestion
let r_loop: int[25] -> int[25] = |st| utils::fold(24, |i| i, st, |acc, r| iota(chi(rho_pi_rearrange(rho_pi_loop(theta_st(acc)))), r) );
```

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -0,0 +1,163 @@
+
+// ln 85 - 86
+// iota
+let iota: int[25], int -> int[25] = |st, r| array::new(24, |i| if i == 0 { xor(st[0], RC[r]) } else { st[i] } );
```
maybe use map_enumerated?

qwang98 replied:
Agreed.

qwang98 commented:
All comments addressed.

A few updates:
1. I created the routine function with all binary function helpers (and, or, xor, not) as parameters in this draft PR (demo only): https://github.com/powdr-labs/powdr/pull/1266/files However, it doesn't look very clean and honestly I think it's better to not have them as parameters, but directly provided in the file. Let me know if you agree.
2. I tried to create a machine out of the asm code I wrote, so that I can refer to the machine in a separate test file. I tried to modify it off the poseidon example, but it's not very intuitive how I should proceed, given that the poseidon example mostly used raw PIL constraints, while I mostly used the asm language (if that's how it's called).

I think I'm at a bottleneck of not knowing how to set my code up for testing, so would be great to have a call on Thursday/Friday about these. I'll reach out separately.

chriseth commented on std/hash/keccaf.asm:
```diff
@@ -9,24 +9,33 @@ let<T1, T2> routine: T1, T1, (T1, T1 -> T1), (T1, T1 -> T1), (T1, T1 -> T1) -> T
     10, 7,  11, 17, 18, 3, 5,  16, 8,  21, 24, 4,
     15, 23, 19, 13, 12, 2, 20, 14, 22, 9,  6,  1
 ];
+// inverse of PI using 1 index plus zero at the beginning
```
Can you add a formal statement about the relation between the two?

chriseth commented:
Yes, the type checker unfortunately does not understand fixed-size arrays.

About testing: I think the proper place should be `pipeline/tests/asm.rs`. The current tests in that file do not do the same thing we should test, but there is a file `powdr_str.rs` in the same directory that calls a function `evaluate_integer_function`. I think a combination of that could allow you to run some tests on individual functions.

Another way of testing could also be to use assertions (`std::check::assert`). You can for example add a `test` function. Or we do it the rust way and add a `test` module inside that file and add helpers that call all functions inside the test module (we don't have annotations unfortunately).

chriseth commented:
I think until everything is working, we should not put this into the std directory, but just into a test file instead.

chriseth commented:
Are the tests running through?

qwang98 commented:
> Are the tests running through?

Yes. I'm able to set up tests. It's just that they don't match existing implementations I'm testing against. I'm trying to debug it now.

qwang98 commented on pil-analyzer/src/evaluator.rs:
```diff
@@ -356,7 +356,7 @@ impl<'a, T: Display> Display for Value<'a, T> {
         match self {
             Value::Bool(b) => write!(f, "{b}"),
-            Value::Integer(x) => write!(f, "{x}"),
+            Value::Integer(x) => write!(f, "{:#x}", x),
```
This is for easier debugging, so I'm leaving it here until we are ready to have it merged.

qwang98 commented:
@chriseth @leonardoalt 

Changes since last time:
1. Passes all test vectors added to the pipeline.
2. Intermediate function tests in commit history, but removed for the review.

Caveat that this is the u64 version, so would still need another adaptation to make it compatible with Goldilocks field.

Questions:
1. Due to the unlimited number of bits integer, extra bitwise and masking is needed in wrapped bit rotations and endianness swaps to confine to 64 bits. Might be good to implement integer types with specific bits in the future.
2. Keccak currently works well alone. Not sure if the next step would be converting it to 32 bits or integrating it with other circuits using connects.

leonardoalt commented:
@qwang98 quick comment: can you fix fmt/lint so the tests run on CI?

chriseth commented:
Oh nice this is great! Can you move the files out of the std directory into a test directory, then we can merge this as a first step.

qwang98 commented:
@chriseth 

Ready for review again! 

The function that we care the most about will be `keccakf_inner`, which takes input after padding and endianness swap.

I also kept test vectors for other functions such as `keccakf` and the full keccak `main`, because they have test vectors that are consistent with Zokrates, so we gain the assurance that the function is implemented correctly and understanding of how `keccakf_inner` is linked to the final keccak result.

chriseth commented on pipeline/tests/asm.rs:
```diff
@@ -454,3 +454,263 @@ fn test_failing_assertion() {
+    }
+
+    // Test keccakf_inner
+    let padded_endianness_swapped_input: Vec<u64> = vec![
```
Why not `= padded_input.iter().map(|x| x.swap_bytes()).collect()`?

chriseth commented on pipeline/tests/asm.rs:
```diff
@@ -454,3 +454,263 @@ fn test_failing_assertion() {
+    ];
+
+    let expected = vec![
+        // ca85d1976d40dcb6ca3becc8c6596e83c0774f4185cf016a05834f5856a37f39
```
What about
```suggestion
        vec!["ca85d1976d40dcb6ca3becc8c6596e83c0774f4185cf016a05834f5856a37f39", ... /* all the others */]
            .into_iter()
            .map(|x|
                (0..x.len())
                    .step_by(2)
                    .map(|i| u8::from_str_radix(&x[i..i + 2], 16).unwrap())
                    .collect()
             ).collect();
```

qwang98 replied:
Nice!

chriseth commented on test_data/asm/keccak.asm:
```diff
@@ -0,0 +1,200 @@
+let not: int -> int = |a| a ^ 0xffffffffffffffff; // bitwise not for 64 bits
+
+// ln 9 - 12
+// TODO: check that these are good for word size 32
```
I think you can remove those

qwang98 replied:
Yes

chriseth commented on test_data/asm/keccak.asm:
```diff
@@ -0,0 +1,200 @@
+    15, 23, 19, 13, 12, 2, 20, 14, 22, 9,  6,  1
+];
+
+// inverse of PI using 1 index plus zero at the beginning
```
Please also add a more formal statement about the relation between PI and PI_INVERSE

qwang98 replied:
Done

chriseth commented on test_data/asm/keccak.asm:
```diff
@@ -0,0 +1,200 @@
+// ln 30 - 33
+// left rotation
+// 32 and 0xffffffff if u32
+let rotl64: int, int -> int = |x, n| and(or((x << n), (x >> (64 - n))), 0xffffffffffffffff); 
```
Shouldn't this move to the beginning of the file? Because I think this is one of the elementary gates.

qwang98 replied:
Agree

chriseth commented on test_data/asm/keccak.asm:
```diff
@@ -0,0 +1,200 @@
+};
+
+// main machine
+// TODO: not sure how to integrate these
```
Please make sure that there are no TODOs left (at least those that are in scope here)

qwang98 replied:
Done
